python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,168 @@
+    deleted from the file system until the last snapshot in which it was
+    listed is garbage collected."""
+
+    snapshot_log: list = Field(alias="snapshot-log")
+    """A list (optional) of timestamp and snapshot ID pairs that encodes
+    changes to the current snapshot for the table. Each time the
+    current-snapshot-id is changed, a new entry should be added with the
+    last-updated-ms and the new current-snapshot-id. When snapshots are
+    expired from the list of valid snapshots, all entries before a snapshot
+    that has expired should be removed."""
+
+    metadata_log: list = Field(alias="metadata-log")
+    """A list (optional) of timestamp and metadata file location pairs that
+    encodes changes to the previous metadata files for the table. Each time
+    a new metadata file is created, a new entry of the previous metadata
+    file location should be added to the list. Tables can be configured to
+    remove oldest metadata log entries and keep a fixed-size log of the most
+    recent entries after a commit."""
+
+    sort_orders: list = Field(alias="sort-orders")
+    """A list of sort orders, stored as full sort order objects."""
+
+    default_sort_order_id: int = Field(alias="default-sort-order-id")
+    """Default sort order id of the table. Note that this could be used by
+    writers, but is not used when reading because reads use the specs stored
+     in manifest files."""
+
+
+class TableMetadataV1(TableMetadataCommonFields, IcebergBaseModel):
+    def __new__(cls, *_, **data):
+        # When we read a V1 format-version, we'll bump it to a V2 table right
+        # away by populating the required fields, and setting the version
+        data["format-version"] = 2
+        schema = data["schema"]
+        if "schemas" not in data:
+            if all([schema != other_schema for other_schema in data["schemas"]]):
+                data["schemas"].append(schema)
+        else:
+            data["schemas"] = [schema]
+        data["current-schema-id"] = schema["schema-id"]
+        data["last-sequence-number"] = _INITIAL_SEQUENCE_NUMBER
+        return TableMetadataV2(**data)
+
+    format_version: Literal[1] = Field(alias="format-version")
+    """An integer version number for the format. Currently, this can be 1 or 2
+    based on the spec. Implementations must throw an exception if a table’s
+    version is higher than the supported version."""
+
+    schema_: Schema = Field(alias="schema")
+    """The table’s current schema. (Deprecated: use schemas and
+    current-schema-id instead)"""
+
+    partition_spec: dict = Field(alias="partition-spec")
+    """The table’s current partition spec, stored as only fields.
+    Note that this is used by writers to partition data, but is
+    not used when reading because reads use the specs stored in
+    manifest files. (Deprecated: use partition-specs and default-spec-id
+    instead)"""
+
+
+class TableMetadataV2(TableMetadataCommonFields, IcebergBaseModel):
```

Comment:
For serializing to JSON, I think it makes sense to have separate classes. I don't see another way to suppress fields like `schema` or `last-sequence-number`.

Reply:
Pydantic will check which version it is, and serialize the fields that are bound to the version. Of course, for V1 the schemas field is optional, and those will be set because we populate them when we initialize a class (the root_validator a comment above).

The following will produce the json with last-sequence-number:
```python
>>> table_metadata = TableMetadataV2(**EXAMPLE_TABLE_METADATA_V2)
>>> table_metadata.json()
{
	"table-uuid": "aefee669-d568-4f9c-b732-3c0cfd3bc7b0",
	"location": "s3://foo/bar/baz.metadata.json",
	"last-updated-ms": 1600000000000,
	"last-column-id": 4,
	"schemas": [{
		"fields": [{
			"id": 1,
			"name": "foo",
			"type": "string",
			"required": true
		}, {
			"id": 2,
			"name": "bar",
			"type": "string",
			"required": true
		}, {
			"id": 3,
			"name": "baz",
			"type": "string",
			"required": true
		}, {
			"id": 4,
			"name": "qux",
			"type": "string",
			"required": true
		}],
		"schema-id": 0,
		"identifier-field-ids": []
	}],
	"current-schema-id": 0,
	"partition-specs": [{
		"spec-id": 0,
		"fields": []
	}],
	"default-spec-id": 0,
	"last-partition-id": 999,
	"properties": {
		"owner": "root",
		"write.format.default": "parquet",
		"read.split.target.size": 134217728
	},
	"current-snapshot-id": 7681945274687743099,
	"snapshots": [{
		"snapshot-id": 7681945274687743099,
		"timestamp-ms": 1637943123188,
		"summary": {
			"operation": "append",
			"added-data-files": "6",
			"added-records": "237993",
			"added-files-size": "3386901",
			"changed-partition-count": "1",
			"total-records": "237993",
			"total-files-size": "3386901",
			"total-data-files": "6",
			"total-delete-files": "0",
			"total-position-deletes": "0",
			"total-equality-deletes": "0"
		},
		"manifest-list": "s3://foo/bar/baz/snap-2874264644797652805-1-9cb3c3cf-5a04-40c1-bdd9-d8d7e38cd8e3.avro",
		"schema-id": 0
	}],
	"snapshot-log": [{
		"timestamp-ms": 1637943123188,
		"snapshot-id": 7681945274687743099
	}],
	"metadata-log": [{
		"timestamp-ms": 1637943123331,
		"metadata-file": "3://foo/bar/baz/00000-907830f8-1a92-4944-965a-ff82c890e912.metadata.json"
	}],
	"sort-orders": [{
		"order-id": 0,
		"fields": []
	}],
	"default-sort-order-id": 0,
	"format-version": 2,
	"last-sequence-number": 1
}
```